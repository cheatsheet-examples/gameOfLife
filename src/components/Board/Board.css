.Board {
    position: relative;
    margin: 0 auto;
    background-color: rgb(18, 56, 224);
    background-image:
        linear-gradient(#333 1px, transparent 1px),
        linear-gradient(90deg, #333 1px, transparent 1px);
}