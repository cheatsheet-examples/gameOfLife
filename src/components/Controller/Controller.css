.controls {
    margin-top: 20px;
}

.button {
    margin-left: 12px;
    cursor: pointer;
}