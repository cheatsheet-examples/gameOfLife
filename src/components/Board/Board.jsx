import React from 'react';
import PropTypes from 'prop-types';
import Cell from '../Cell/Cell';
import './Board.css';

const Board = ({ handleSingleClick, width, height, cellSize, cells, boardRef }) => (

  <div className="Board"
    style={{ width: width, height: height, backgroundSize: `${cellSize}px ${cellSize}px` }}
    onClick={handleSingleClick}
    ref={boardRef}
  >
    {cells.map(cell => (
      <Cell key={`${cell.x},${cell.y}`} x={cell.x} y={cell.y} cellSize={cellSize} />
    ))}
  </div>
);

Board.propTypes = {
  handleSingleClick: PropTypes.func.isRequired,
  width: PropTypes.number,
  height: PropTypes.number,
  cellSize: PropTypes.number,
  cells: PropTypes.array.isRequired,
  boardRef: PropTypes.func,
};

Board.defaultProps = {
  cells: [],
};

export default Board;