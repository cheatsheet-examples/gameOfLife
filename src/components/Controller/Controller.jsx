import React from 'react';
import './Controller.css';
import PropTypes from 'prop-types';

const Controller = ({ stopGame, playGame, generateRandomBoard, clearCells,
  isGameRunning }) => (
    <div className="controls">

      {isGameRunning ?
        <button className="button" onClick={stopGame}>Stop</button> :
        <button className="button" onClick={playGame}>Play</button>
      }
      <button className="button" onClick={generateRandomBoard}>Random</button>
      <button className="button" onClick={clearCells}>Clear</button>

    </div>
  );

Controller.propTypes = {
  stopGame: PropTypes.func.isRequired,
  playGame: PropTypes.func.isRequired,
  generateRandomBoard: PropTypes.func.isRequired,
  clearCells: PropTypes.func.isRequired,
  isGameRunning: PropTypes.bool.isRequired
};

Controller.defaultProps = {
  isGameRunning: false
};
export default Controller;