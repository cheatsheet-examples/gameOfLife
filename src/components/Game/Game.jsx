import React, { Component } from 'react';
import Broad from '../Board/Board';
import Controller from '../Controller/Controller';

const WIDTH = 800;
const HEIGHT = 300;
const CELL_SIZE = 20;
const INTERVAL = 400;
const RANDON_RATE = 0.5;

class Game extends Component {
    constructor() {
        super();
        this.rows = HEIGHT / CELL_SIZE;
        this.cols = WIDTH / CELL_SIZE;
        this.board = this.generateEmptyBoard();
    }

    state = {
        cells: [],
        isGameRunning: false,
    }

    generateEmptyBoard() {
        return Array.from(Array(this.rows), () => new Array(this.cols).fill(false));
    }

    generateCells() {
        let cells = [];
        this.board.forEach((cols, y) => {
            cols.forEach((col, x) => {
                if (col) {
                    cells.push({ x, y });
                }
            })
        });
        return cells;
    }

    handleSingleClick = (event) => {
        this.stopGame();
        const elemOffset = this.getBoardBound();
        const offsetX = event.clientX - elemOffset.x;
        const offsetY = event.clientY - elemOffset.y;
        const x = Math.floor(offsetX / CELL_SIZE);
        const y = Math.floor(offsetY / CELL_SIZE);

        if (x >= 0 && x <= this.cols && y >= 0 && y <= this.rows) {
            this.board[y][x] = !this.board[y][x];
            this.setState({ cells: this.generateCells() });
        }
    }

    playGame = () => {
        this.setState({ isGameRunning: true });
        this.runNextStep();
    }

    stopGame = () => {
        this.setState({ isGameRunning: false });
        if (this.timeoutHandler) {
            window.clearTimeout(this.timeoutHandler);
            this.timeoutHandler = null;
        }
    }

    getBoardBound() {
        return this.boardRef.getBoundingClientRect();
    }

    runNextStep() {
        this.goNextStep();
        this.timeoutHandler = window.setTimeout(() => {
            this.runNextStep();
        }, INTERVAL);
    }

    goNextStep() {
        let newBoard = this.generateEmptyBoard();
        let oldBoard = this.board;
        this.board.forEach((cols, y) => {
            cols.forEach((col, x) => {
                let neighbors = this.getNeighbors(oldBoard, x, y);
                if (oldBoard[y][x]) {
                    newBoard[y][x] = (neighbors === 2 || neighbors === 3) ? true : false;

                } else {
                    if (!oldBoard[y][x] && neighbors === 3) {
                        newBoard[y][x] = true;
                    }
                }
            })
        });

        this.board = newBoard;
        this.setState({ cells: this.generateCells() });
    }

    getNeighbors(board, x, y) {

        let neighbors = 0;
        const dirs = [[-1, -1], [-1, 0], [-1, 1], [0, 1], [1, 1], [1, 0], [1, -1], [0, -1]];
        dirs.forEach((dir) => {
            let y1 = y + dir[0];
            let x1 = x + dir[1];

            if (x1 >= 0 && x1 < this.cols && y1 >= 0 && y1 < this.rows) {
                if (board[y1][x1]) {
                    neighbors++;
                }
            }
        });
        return neighbors;
    }


    clearCells = () => {
        this.stopGame();
        this.board = this.generateEmptyBoard();
        this.setState({ cells: [] });
    }

    generateRandomBoard = () => {
        this.stopGame();
        this.board.forEach((cols, y) => {
            cols.forEach((col, x) => {
                this.board[y][x] = (Math.random() >= RANDON_RATE);
            })
        });
        this.setState({ cells: this.generateCells() });
    }

    render() {
        const { cells, isGameRunning } = this.state;
        return (
            <div>
                <h1>Game of Life</h1>
                <Broad handleSingleClick={this.handleSingleClick} width={WIDTH} height={HEIGHT}
                    cellSize={CELL_SIZE} cells={cells} boardRef={(n) => { this.boardRef = n; }} />
                <Controller stopGame={this.stopGame} playGame={this.playGame} generateRandomBoard={this.generateRandomBoard}
                    clearCells={this.clearCells}
                    isGameRunning={isGameRunning}
                />
            </div>
        );
    }
}
export default Game;
