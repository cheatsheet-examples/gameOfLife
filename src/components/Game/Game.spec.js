import React from 'react';
import { shallow } from 'enzyme';
import Game from './Game';
import Board from '../Board/Board';
import Controller from '../Controller/Controller';

describe('Game', () => {
    let wrapper;
    beforeEach(() => (wrapper = shallow(<Game />)));

    it('should render correctly', () => expect(wrapper).toMatchSnapshot());

    it('should render a <div />', () => {
        expect(wrapper.find('div').length).toEqual(1);
    });

    it('should render the Board and Controller Components', () => {
        expect(
            wrapper.containsAllMatchingElements([
                <Board
                    handleSingleClick={wrapper.instance().handleSingleClick}
                    cells={wrapper.instance().state.cells} />,
                <Controller
                    stopGame={wrapper.instance().stopGame}
                    playGame={wrapper.instance().playGame}
                    generateRandomBoard={wrapper.instance().generateRandomBoard}
                    isGameRunning={wrapper.instance().state.isGameRunning}
                    clearCells={wrapper.instance().clearCells}
                />,
            ])
        ).toEqual(true);

    });
    it('should get empty board', () => {
        wrapper.instance().rows = 3;
        wrapper.instance().cols = 3;
        let mockBoard = [
            [false, false, false],
            [false, false, false],
            [false, false, false],
        ];
        let expectEmptyBoard = wrapper.instance().generateEmptyBoard();
        expect(expectEmptyBoard).toEqual(mockBoard);
    });

    it('should generate empty cells array by empty board', () => {
        wrapper.instance().rows = 3;
        wrapper.instance().cols = 3;
        wrapper.instance().generateEmptyBoard();
        let expectCells = wrapper.instance().generateCells();
        expect(expectCells).toEqual([]);

    });

    it('should generate cells array by board have true', () => {
        wrapper.instance().board = [
            [true, false, false],
            [false, false, false],
            [false, false, true],
        ];

        let expectCells = wrapper.instance().generateCells();
        expect(expectCells.length).toEqual(2);
        expect(expectCells).toEqual([{ x: 0, y: 0 }, { x: 2, y: 2 }]);
    });

    it('should random generate cells could not be all be true or false', () => {
        wrapper.instance().generateRandomBoard();
        expect(wrapper.state('cells')).not.toEqual([]);
        expect(wrapper.state('cells').length < wrapper.instance().cols * wrapper.instance().rows).toBeTruthy();

    });
    describe('calculate next step of game of life', () => {
        it('should still live if centre cell have 2 neighbors', () => {
            wrapper.instance().board = [
                [true, false, false],
                [false, true, false],
                [false, false, true],
            ];
            wrapper.instance().rows = 3;
            wrapper.instance().cols = 3;

            let expectCells = wrapper.instance().generateCells();
            expect(expectCells.length).toEqual(3);
            wrapper.instance().goNextStep();
            let actualBoard = [
                [false, false, false],
                [false, true, false],
                [false, false, false],
            ];
            expect(wrapper.state('cells')).toEqual([{ x: 1, y: 1 }]);
            expect(wrapper.instance().board).toEqual(actualBoard);
        });
        it('should still live if centre cell have 3 neighbors', () => {
            wrapper.instance().board = [
                [true, false, false],
                [false, true, true],
                [false, false, false],
            ];
            wrapper.instance().rows = 3;
            wrapper.instance().cols = 3;
            wrapper.instance().goNextStep();
            expect(wrapper.state('cells')).toEqual([{ x: 1, y: 0 }, { x: 1, y: 1 }]);
            expect(wrapper.instance().board[1][1]).toEqual(true);
        });
        it('should still live if dead centre cell have 3 neighbors', () => {
            wrapper.instance().board = [
                [true, false, false],
                [false, false, true],
                [false, true, false],
            ];
            wrapper.instance().rows = 3;
            wrapper.instance().cols = 3;

            wrapper.instance().goNextStep();

            expect(wrapper.state('cells')).toEqual([{ x: 1, y: 1 }]);
            expect(wrapper.instance().board[1][1]).toEqual(true);
        });
        it('should  dead if 1 neighbors', () => {
            wrapper.instance().board = [
                [true, false, false],
                [false, true, false],
                [false, false, false],
            ];
            wrapper.instance().rows = 3;
            wrapper.instance().cols = 3;

            wrapper.instance().goNextStep();
            expect(wrapper.state('cells')).toEqual([]);
            expect(wrapper.instance().board[1][1]).toEqual(false);
        });
        it('should  dead if 4 neighbors', () => {
            wrapper.instance().board = [
                [true, false, true],
                [false, true, false],
                [true, false, true],
            ];
            wrapper.instance().rows = 3;
            wrapper.instance().cols = 3;

            wrapper.instance().goNextStep();
            expect(wrapper.instance().board[1][1]).toEqual(false);
        });
    });

    describe('check neighbors number ', () => {
        it('should get 0 neighbor true when board all cells are flase', () => {
            let mockBoard = [
                [false, false, false],
                [false, false, false],
                [false, false, false],
            ];
            const expectNumber = wrapper.instance().getNeighbors(mockBoard, 1, 1);
            expect(expectNumber).toEqual(0);
        });

        it('should get 1 neighbor true when board has [0,0] true', () => {
            let mockBoard = [
                [true, false, false],
                [false, false, false],
                [false, false, false],
            ];
            const expectNumber = wrapper.instance().getNeighbors(mockBoard, 1, 1);
            expect(expectNumber).toEqual(1);
        });
        it('should get 2 neighbor true when board has [0,0] and [2,2] true', () => {
            let mockBoard = [
                [true, false, false],
                [false, false, false],
                [false, false, true],
            ];
            const expectNumber = wrapper.instance().getNeighbors(mockBoard, 1, 1);
            expect(expectNumber).toEqual(2);
        });
        it('should get 8 neighbor true when middle cell all neighbors are true', () => {
            let mockBoard = [
                [true, true, true],
                [true, false, true],
                [true, true, true],
            ];
            const expectNumber = wrapper.instance().getNeighbors(mockBoard, 1, 1);
            expect(expectNumber).toEqual(8);
        });
    });
    describe('button click', () => {
        it('isGameRuning status change for play game', () => {
            wrapper.instance().playGame();
            expect(wrapper.state('isGameRunning')).toEqual(true);
            expect(wrapper.instance().timeoutHandler).not.toBeNull();
        });
        it('isGameRuning status change for stop game', () => {
            wrapper.instance().stopGame();
            expect(wrapper.state('isGameRunning')).toEqual(false);
            expect(wrapper.instance().timeoutHandler).toBeUndefined();
        });
        it('check state after clear cells', () => {
            wrapper.instance().clearCells();
            expect(wrapper.state('isGameRunning')).toEqual(false);
            expect(wrapper.instance().timeoutHandler).toBeUndefined();
            expect(wrapper.state('cells')).toEqual([]);
        });
        it('check state after generate random', () => {
            wrapper.instance().generateRandomBoard();
            expect(wrapper.state('isGameRunning')).toEqual(false);
            expect(wrapper.instance().timeoutHandler).toBeUndefined();
            expect(wrapper.state('cells')).not.toEqual([]);
        });
        it('check state after click on the board', () => {
            const mockElemOffset = jest.spyOn(wrapper.instance(), 'getBoardBound');
            mockElemOffset.mockReturnValue({ x: 0, y: 0 });
            wrapper.instance().handleSingleClick({ clientX: 15, clientY: 15 });
            expect(wrapper.state('isGameRunning')).toEqual(false);
            expect(wrapper.state('cells')).toEqual([{ x: 0, y: 0 }]);
            expect(wrapper.instance().board[0, 0]).toBeTruthy();


            wrapper.instance().handleSingleClick({ clientX: 30, clientY: 30 });
            expect(wrapper.state('cells')).toEqual([{ x: 0, y: 0 }, { x: 1, y: 1 }]);
            expect(wrapper.instance().board[1, 1]).toBeTruthy();

            wrapper.instance().handleSingleClick({ clientX: 15, clientY: 15 });
            expect(wrapper.state('cells')).toEqual([{ x: 1, y: 1 }]);

        });
    });
});