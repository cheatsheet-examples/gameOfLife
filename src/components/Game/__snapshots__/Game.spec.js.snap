// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render correctly 1`] = `
<div>
  <h1>
    Game of Life
  </h1>
  <Board
    boardRef={[Function]}
    cellSize={20}
    cells={Array []}
    handleSingleClick={[Function]}
    height={300}
    width={800}
  />
  <Controller
    clearCells={[Function]}
    generateRandomBoard={[Function]}
    isGameRunning={false}
    playGame={[Function]}
    stopGame={[Function]}
  />
</div>
`;
